MSAL/src/requests/MSALBaseRequest.m

```diff
@@ -195,52 +195,111 @@ - (void)acquireToken:(nonnull MSALCompletionBlock)completionBlock
-            completionBlock(nil, userMismatchError);
+            completionBlock(nil, scopesError);
             return;
         }
-        
-        MSALResult *result = [MSALResult resultWithAccessToken:accessToken idToken:idToken isExtendedLifetimeToken:NO];
-        
-        completionBlock(result, nil);
+
+        NSError *resultError = nil;
+        MSALResult *result = [self resultFromTokenResponse:tokenResponse error:&resultError];
+        completionBlock(result, resultError);
     }];
 }
 
+- (BOOL)verifyTokenResponse:(MSIDAADV2TokenResponse *)tokenResponse
+                      error:(NSError **)error
+{
+    NSError *msidError = nil;
+
+    if (![self.oauth2Factory verifyResponse:tokenResponse context:nil error:&msidError])
+    {
+        if (error) *error = [MSALErrorConverter MSALErrorFromMSIDError:msidError];
+        return NO;
+    }
+
+    if (_parameters.account.homeAccountId.identifier != nil &&
+        ![_parameters.account.homeAccountId.identifier isEqualToString:tokenResponse.clientInfo.accountIdentifier])
+    {
+        NSError *userMismatchError = CREATE_MSAL_LOG_ERROR(_parameters, MSALErrorMismatchedUser, @"Different user was returned from the server");
+        if (error) *error = userMismatchError;
+
+        return NO;
+    }
+
+    return YES;
+}
+
+- (MSALResult *)resultFromTokenResponse:(MSIDAADV2TokenResponse *)tokenResponse
+                                  error:(NSError **)error
+{
+    MSIDAccessToken *accessToken = [self.oauth2Factory accessTokenFromResponse:tokenResponse configuration:_parameters.msidConfiguration];
+    MSIDIdToken *idToken = [self.oauth2Factory idTokenFromResponse:tokenResponse configuration:_parameters.msidConfiguration];
+
+    if (!accessToken || !idToken)
+    {
+        NSError *resultError = CREATE_MSAL_LOG_ERROR(_parameters, MSALErrorBadTokenResponse, @"Bad token response returned from the server");
+
+        if (error) *error = resultError;
+    }
+
+    MSALResult *result = [MSALResult resultWithAccessToken:accessToken idToken:idToken isExtendedLifetimeToken:NO];
+    return result;
+}
+
+- (BOOL)verifyScopesWithResponse:(MSIDAADV2TokenResponse *)tokenResponse
+                           error:(NSError **)error
+{
+    /*
+        If server returns less scopes than developer requested,
+        we'd like to throw an error and specify which scopes were granted and which ones not
+     */
+
+    NSOrderedSet *grantedScopes = [tokenResponse.scope scopeSet];
+    MSIDConfiguration *configuration = _parameters.msidConfiguration;
+
+    if (![configuration.scopes isSubsetOfOrderedSet:grantedScopes])
+    {
+        if (error)
+        {
+            NSMutableDictionary *additionalUserInfo = [NSMutableDictionary new];
+            additionalUserInfo[MSALGrantedScopesKey] = [grantedScopes array];
+
+            NSMutableOrderedSet *requestedScopeSet = [configuration.scopes mutableCopy];
```

Comment:
nit: declinedScopeSet = [configuration.scopes mutableCopy] minusOrderedSet... ?
That way, next line would read easier.

Reply:
agree, fixed